### PR TITLE
Added mypy pre-commit hook

### PR DESCRIPTION
**Pull request by arnav13081994**
## Description

Added `mypy` pre-commit hook

Checklist:

- [X] I've made sure that `tests/test_cookiecutter_generation.py` is updated accordingly (especially if adding or updating a template option)
- [X] I've updated the documentation or confirm that my change doesn't require any updates

## Rationale

Adding `mypy` as another pre-commit hook will have the following benefits:

1) Compile time errors can be discovered while developing locally.
1) Compile time errors can be discovered faster

### CODE REVIEW

**Comment by foarsitter** [comment]
I'm against this change since `mypy` cannot fix the problems it encounters. Some typing problems are hard to solve and this would block the user from commiting which would disadvantage beginners.

File: {{cookiecutter.project_slug}}/mypy.sh
Comment (browniebroke) [inline {{cookiecutter.project_slug}}/mypy.sh:1]: That seems a bit too complicated to me to maintain... It's a bit too beyond my bash skills 😄 

How should folks run this hook? Should it run inside of the Docker container? If it doesn't work outside of Docker than it's a blocker.

File: {{cookiecutter.project_slug}}/.github/workflows/ci.yml [hunk 1]
```diff
@@ -28,6 +28,11 @@ jobs:
         with:
           python-version: 3.9
 
+      - name: Install All Project Dependencies for pylint and mypy to check
+        run: |
+          python -m pip install --upgrade pip
+          pip install -r requirements/local.txt
```
Comment (browniebroke) [inline {{cookiecutter.project_slug}}/.github/workflows/ci.yml:34]: That seems like a big overhead

**Review by browniebroke — COMMENTED** [review]
I'm also against this change. I personally think that it could be nice to run mypy on pre-commit (in terms of UX).

Unfortunately, the fact that it needs the project dependencies in a blocker in my opinion.

**Pull request closed without merging**